frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1049 @@
+
+		/// What to do with initial voting members of the Alliance.
+		type InitializeMembers: InitializeMembers<Self::AccountId>;
+
+		/// What to do when a member has been added or removed.
+		type MembershipChanged: ChangeMembers<Self::AccountId>;
+
+		/// The identity verifier of an Alliance member.
+		type IdentityVerifier: IdentityVerifier<Self::AccountId>;
+
+		/// The provider of the proposal operation.
+		type ProposalProvider: ProposalProvider<Self::AccountId, Self::Hash, Self::Proposal>;
+
+		/// Maximum number of proposals allowed to be active in parallel.
+		type MaxProposals: Get<ProposalIndex>;
+
+		/// The maximum number of founders supported by the pallet. Used for weight estimation.
+		///
+		/// NOTE:
+		/// + Benchmarks will need to be re-run and weights adjusted if this changes.
+		/// + This pallet assumes that dependencies keep to the limit without enforcing it.
+		type MaxFounders: Get<u32>;
+
+		/// The maximum number of fellows supported by the pallet. Used for weight estimation.
+		///
+		/// NOTE:
+		/// + Benchmarks will need to be re-run and weights adjusted if this changes.
+		/// + This pallet assumes that dependencies keep to the limit without enforcing it.
+		type MaxFellows: Get<u32>;
+
+		/// The maximum number of allies supported by the pallet. Used for weight estimation.
+		///
+		/// NOTE:
+		/// + Benchmarks will need to be re-run and weights adjusted if this changes.
+		/// + This pallet assumes that dependencies keep to the limit without enforcing it.
+		type MaxAllies: Get<u32>;
+
+		/// The maximum length of the blacklist supported by the pallet.
+		#[pallet::constant]
+		type MaxBlacklistCount: Get<u32>;
+
+		/// The maximum length of a website URL.
+		#[pallet::constant]
+		type MaxWebsiteUrlLength: Get<u32>;
+
+		/// The deposit required for submitting candidacy.
+		#[pallet::constant]
+		type AllyDeposit: Get<BalanceOf<Self, I>>;
+
+		/// The maximum number of announcements.
+		#[pallet::constant]
+		type MaxAnnouncementsCount: Get<u32>;
+
+		/// The maximum number of members per member role. Should not exceed the sum of
+		/// `MaxFounders` and `MaxFellows`.
+		#[pallet::constant]
+		type MaxMembersCount: Get<u32>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// The founders/fellows/allies have already been initialized.
+		MembersAlreadyInitialized,
+		/// Account is already a member.
+		AlreadyMember,
+		/// Account is not a member.
+		NotMember,
+		/// Account is not an ally.
+		NotAlly,
+		/// Account is not a founder.
+		NotFounder,
+		/// This member is up for being kicked from the Alliance and cannot perform this operation.
+		UpForKicking,
+		/// Account does not have voting rights.
+		NoVotingRights,
+		/// Account is already an elevated (fellow) member.
+		AlreadyElevated,
+		/// Item is already in the blacklist.
+		AlreadyInBlacklist,
+		/// Item is not blacklisted.
+		NotInBlacklist,
```

Comment:
Should we have NotInBlacklist(usize) where we specify which item's index was offending in the input? Would take out some of the guesswork when they have to re-submit without that item.

Reply:
Seems reasonable, but it does fail on the first one, so there could be further items.